llvm/test/CodeGen/AArch64/mlicm-csr-mask.mir

```diff
@@ -0,0 +1,49 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=aarch64-unknown-linux-gnu -run-pass=greedy,machinelicm -verify-machineinstrs -debug -o - %s | FileCheck %s
+
+# FIXME: Running RA is needed otherwise it runs pre-RA LICM.
+---
+name: test
+tracksRegLiveness: true
+body: |
+  ; CHECK-LABEL: name: test
+  ; CHECK: bb.0:
+  ; CHECK-NEXT:   successors: %bb.1(0x80000000)
+  ; CHECK-NEXT:   liveins: $x0, $w1, $x2
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   B %bb.1
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT: bb.1:
+  ; CHECK-NEXT:   successors: %bb.1(0x40000000), %bb.2(0x40000000)
+  ; CHECK-NEXT:   liveins: $x0, $w1, $x2
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   renamable $q11 = MOVIv4i32 2, 8
+  ; CHECK-NEXT:   BL &memset, csr_aarch64_aapcs, implicit-def dead $lr, implicit $sp, implicit $x0, implicit $w1, implicit $x2, implicit-def $sp, implicit-def $x0
+  ; CHECK-NEXT:   renamable $q10 = MVNIv4i32 4, 0
```

Comment:
What does this test actually demonstrate? Without this patch, I see that the def of `q10` gets hoisted past this memset call:
```
BL &memset, <regmask $fp $lr $wzr $xzr $b8 $b9 $b10 $b11 $b12 $b13 $b14 $b15 $d8 $d9 $d10 $d11 $d12 $d13 $d14 $d15 $h8 $h9 $h10 $h11 
$h12 $h13 $h14 $h15 $s8 $s9 $s10 $s11 $s12 $s13 $s14 $s15 $w19 $w20 $w21 $w22 $w23 $w24 $w25 $w26 $w27 $w28 $w29 $w30 $x19 $x20 $x21 $x22 $x23 
$x24 $x25 $x26 $x27 $x28 $d8_d9 $d9_d10 $d10_d11 $d11_d12 $d12_d13 $d13_d14 $d14_d15 $d8_d9_d10_d11 $d9_d10_d11_d12 $d10_d11_d12_d13 $d11_d12_d
13_d14 $d12_d13_d14_d15 $d8_d9_d10 $d9_d10_d11 $d10_d11_d12 $d11_d12_d13 $d12_d13_d14 $d13_d14_d15 $x22_x23_x24_x25_x26_x27_x28_fp $x20_x21_x22
_x23_x24_x25_x26_x27 $w20_w21 $w22_w23 $w24_w25 $w26_w27 $w28_w29 $x28_fp $x20_x21 $x22_x23 $x24_x25 $x26_x27>, implicit-def dead $lr, implicit
 $sp, implicit $x0, implicit $w1, implicit $x2, implicit-def $sp, implicit-def $x0
```
I don't understand why this happened. `q10` has subregs `d20` and `d21` but I don't see any of those listed in this regmask.

Reply:
D'oh - I thought this test case _was_ ARM! Let me take another look...

Reply:
I was trying to get my head around what was going on. There is a rule in the AArch64 PCS which talks about the top bits of a Q register needing to be preserved across a call:
```
Registers v8-v15 must be preserved by a callee across subroutine calls; the remaining registers (v0-v7,
v16-v31) do not need to be preserved (or should be preserved by the caller). Additionally, only the bottom 64
bits of each value stored in v8-v15 need to be preserved 7; it is the responsibility of the caller to preserve
larger values.
```

Reply:
OK so AArch64 registers work like this: `b10` aliases the low 8 bits of `h10` which aliases the low 16 bits of `s10` which aliases the low 32 bits of `d10` which aliases the low 64 bits of `q10` which aliases the low 128 bits of `z10`.

My suspicion is that TableGen only creates one regunit for all of these `*10` registers. If that's true then there's no way for regunit-based clobbering info to express that `d10` is preserved but `q10` is not (because its high bits are clobbered). So I think we would have to go back and revert #94608.

Reply:
There should be regunits created for each trackable, even if non-addressable parts of the register.

Reply:
I agree that would be nice but that is not what TableGen currently does. I've just checked and all of b10 h10 s10 d10 q10 z10 have a single regunit number 35.

Reply:
> My suspicion is that TableGen only creates one regunit for all of these *10 registers.

Yes, I think it's what's happening too. It's why I couldn't reconcile the AMDGPU case with this one.

> So I think we would have to go back and revert https://github.com/llvm/llvm-project/pull/94608.

I would rather add some system to let MachineLICM work with either RUs or RegAliases through some small additional abstractions than do a full revert. The old approach uses RegAliases way too much and it's way too expensive on AMDGPU.

> There should be regunits created for each trackable, even if non-addressable parts of the register.

Would this be a lot of work? It'd be a much neater fix than having to work-around this.

Reply:
We could perhaps do a targeted fix for AArch64 too, and create an artificial "high" register for Q registers to model this.
@davemgreen What do you think about that?

To recap, we have 3 options:
- revert the MachineLICM change
   - As the change is too good for AMDGPU, I would instead implement the 2 approaches side-by-side and use TRI to switch between both. All targets would use the new approach except AArch.
- Fix regunits calculations
  - Perfect fix in theory, in practice I feel like this could have a big impact on regalloc as a whole so I'm pessimistic it'd be a small change that lands quickly
- Add fake high registers for AArch64 Q registers
  - I think we do that for some registers on AMDGPU already. On one hand it can be seen as a the right thing to do to model high bits properly, OTOH it's a hack around shortcomings of the register modeling we have now in LLVM.

Option 1 vs Option 3 is basically a decision or whether we want the workaround to be in the backend, or in the pass. I think option 2 needs to be done anyway but I expect it'll be a longer task.

Reply:
>  If you declare a register with two subregs, tablegen does not know (in general) whether those subregs completely cover the register

This does not spark joy

> Yes, I wondered about that, but it makes me a little nervous that it might leave latent bugs on other targets.

I think with how regunitts work, anything that uses regunits could have similar bugs.

I'm wondering if we could implement some simple logic like this:
- Add a flag to registers to tell TableGen that subregs cover the whole register
- If a register has subregs, and:
  - The flag is set: we generate regunits as we do currently.
  - The flag is not set: we generate one additional regunit exclusive for that register to represent the unknown bits of that registers.

So in the case of AArch64, for Qn registers we'd have n more regunits to represent the upper unknown bits.

Reply:
> Add a flag to registers to tell TableGen that subregs cover the whole register (EDIT: we already have one)

What is that flag that we already have?

Anyway I think this kind of approach sounds reasonable, but should probably get some wider discussion. Possible alternative ideas:
- Declare the width of every physical register, so that tablegen can work out by itself whether or not it is completely covered by subregs.
- Change the AArch64 register definitions to declare some kind of dummy subreg for the high parts, which is only used to force tablegen to generate the extra regunits. The only advantage of this is that we might not have to modify tablegen at all.

Reply:
Hello. Sorry, all the subtleties of RegUnits are not something I know a lot about (and I may be a bit unresponsive today, sorry if I am). Perhaps @efriedma-quic or @TNorthover know more about how it should all work. We would usually revert back to a known-good state or fix quickly, and leaving it broken for any length of time isn't the llvm way of doing things. So I'm all in favour of this fix, thanks.

We have scalable registers in AArch64 aliasing the same registers too, to throw another potential spanner in the works. I don't believe those registers actually get marked as scalable at the moment though. If we go with option 3 then we should try and make sure that works OK too.

Reply:
Sorry all, I'm just seeing this thread now.
Apologies if I repeat things that have already been said, I haven't read through the whole history.


> I agree that would be nice but that is not what TableGen currently does. I've just checked and all of b10 h10 s10 d10 q10 z10 have a single regunit number 35.

@jayfoad is right this is how the regunits work and the issue with register mask is well known (well at least @MatzeB and I know :P). We shouldn't use regunits for regmask operands because they are indeed not precise enough until we go with option 2. 

(And by the way changing regunits shouldn't change regalloc. I'm curious to see why you @Pierre-vh saw this.)